### DISCUSSION TITLE
Uploading wheels silently fail

### DISCUSSION THREAD

**Discussion by DataTriny** [PyO3/maturin#1721] [Q&A]
Hello everyone,

I am trying to test the publishing of a crate built with PyO3 using Maturin on Linux.

So I first spin a Docker container by running this inside the root directory of my Cargo workspace:

```bash
docker run --rm -it -v $(pwd):/io --entrypoint /bin/bash ghcr.io/pyo3/maturin
```

I then `cd` into the directory that contains my crate, and run:

```bash
maturin publish --repository-url https://test.pypi.org/simple --username USERNAME --password PASSWORD
```

Which outputs:

```
🔗 Found pyo3 bindings with abi3 support for Python ≥ 3.7
🐍 Not using a specific python interpreter
📡 Using build options features from pyproject.toml
    Finished release [optimized + debuginfo] target(s) in 0.05s
📦 Built wheel for abi3 Python ≥ 3.7 to /io/target/wheels/accesskit-0.1.0-cp37-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
📦 Built source distribution to /io/target/wheels/accesskit-0.1.0.tar.gz
🚀 Uploading 2 packages
✨ Packages uploaded successfully
```

However, the project never showed up on `test.pypi.org`. In fact, I can run this command multiple times and it always succeeds. I've tried replacing my password by an access token, but the website shows it is never used.

Am I missing something so obvious it is not documented anywhere?

maturin 1.2.0

[accesskit-0.1.0-cp37-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.zip](https://github.com/PyO3/maturin/files/12294701/accesskit-0.1.0-cp37-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.zip)

Thanks in advance for your help.

**Comment by DataTriny** [comment]
If I change the repository URL to `https://test.pypi.org/legacy/`, the upload succeeds for real. Tokens seem to not be supported by this endpoint though.
I should also note that I initially removed the trailing slash after `/simple` because I otherwise get a 405 error. It seems that what is advertised on the README in the "Python packaging basics" section does not work in my case.

**Comment by DataTriny** [comment]
OK now I understand my mistake. The phrasing in the README led me to think that the endpoints for downloading packages and uploading packages were the same, but it's indeed not the case. Maybe this section could include the command to publish to `test.pypi.org`?

**Reply by messense** [reply]
You can use `maturin upload --repository testpypi` instead of specifying the full upload endpoint URL.
